docs/developer-guide/custom-widgets/universal-widgets.md

```diff
@@ -0,0 +1,339 @@
+
+```ts
+class AwesomeWidget implements Widget {
+  onRemove() {
+    // Cleanup code
+  }
+}
+```
+
+-----
+
+## Handling View Interaction Events
+
+A widget can update in response to a user interacting with the deck.gl view the widget is attached to. 
+
+```ts
+class AwesomeWidget implements Widget {
+  viewId = 'minimap'
+
+  onClick(info, event) {
+    // Called when a click event occurs in the minimap view.
+  }
+}
+```
+
+See [`Methods`](../../api-reference/core/widget.md#methods) for a complete list.
+
+## Customizing Styles
+
+A good universal widget provides users with ways to customize the styles of any elements it creates. Three common ways to do with are:
+
+- Inline styles prop
+- CSS class prop
+- CSS variables
+
+### Inline Styles Prop
+
+Define a prop for overriding styles inline
+
+```ts
+interface AwesomeWidgetProps {
+  style?: Partial<CSSStyleDeclaration>;
+}
+```
+
+Apply the styles to your widget
+
+```ts
+import {
+  type Widget,
+  _deepEqual as deepEqual,
+  _applyStyles as applyStyles,
+  _removeStyles as removeStyles
+} from '@deck.gl/core'
+
+class AwesomeWidget implements Widget {
+
+  constructor(props: AwesomeWidgetProps) {
+    ...
+    this.props = {
+      ...props,
+      style: props.style || {}
+    }
+  }
+
+  onAdd() {
+    const {style} = this.props;
+    const element = document.createElement('div');
+    applyStyles(element, style);
+    ...
+  }
+  
+  setProps(props: Partial<AwesomeWidgetProps>) { 
+    const el = this.element;
+    if (el) {
+      if (!deepEqual(oldProps.style, props.style, 1)) {
+        removeStyles(el, oldProps.style);
+        applyStyles(el, props.style);
+      }
+    }
+    ...
+  }
+
+}
+```
+
+### CSS Class Prop
+
+Define a prop for adding a CSS class
+
+```ts
+interface AwesomeWidgetProps {
+  /**
+   * Additional CSS class.
+   */
+  className?: string;
+}
+```
+
+Apply the CSS class to your widget
+
+```ts
+import { type Widget } from '@deck.gl/core'
+
+class AwesomeWidget implements Widget {
+
+  onAdd() {
+    const {className} = this.props;
+    const element = document.createElement('div');
+    if (className) element.classList.add(className);
+    ...
+  }
+  
+  setProps(props: Partial<AwesomeWidgetProps>) {
+    const oldProps = this.props;
+    const el = this.element;
+    if (el) {
+      if (oldProps.className !== props.className) {
+        if (oldProps.className) el.classList.remove(oldProps.className);
+        if (props.className) el.classList.add(props.className);
+      }
+    }
+    ...
+  }
+}
+```
+
+### CSS Variables
+
+Define variables in the widget's CSS stylesheet
+
+```css
+.my-awesome-widget {
+  background-color: var(--primary-color, rebeccapurple);
+}
+```
+
+Override in user's application
+
+```css
+.my-awesome-widget {
+    --primary-color: chartreuse;
+}
+```
+
+## Example: Layer List Widget with Preact
+
+Below is a comprehensive example demonstrating a layer list widget implemented using Preact for dynamic UI rendering:
+
+```tsx
```

Comment:
We should really use MDX capabilities and make these widgets live on the pages...